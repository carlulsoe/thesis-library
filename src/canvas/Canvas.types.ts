export interface CanvasProps {
  setup: (ctx: CanvasRenderingContext2D) => void;
  width: number;
  height: number;
}
