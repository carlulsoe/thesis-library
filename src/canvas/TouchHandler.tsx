type SimplifiedTouch = {
  x: number;
  y: number;
  w: number;
  h: number;
};

type SimplifiedTouchEvent = {
  ctrlKey: boolean;
  touches: SimplifiedTouch[];
};

function simplifyTouchEvent(event: TouchEvent) {
  let touchList: SimplifiedTouch[] = [];
  for (let i = 0; i < event.touches.length; i++) {
    let item = event.touches.item(i)!;
    touchList[i] = {
      w: item.radiusX,
      h: item.radiusY,
      x: item.clientX - item.radiusX / 2,
      y: item.clientY - item.radiusY / 2,
    };
  }

  const simplifiedTouchEvent: SimplifiedTouchEvent = {
    ctrlKey: event.ctrlKey,
    touches: touchList,
  };
  return simplifiedTouchEvent;
}

function draw(touch: SimplifiedTouch, context: CanvasRenderingContext2D) {
  context.beginPath();
  context.arc(
    touch.x + touch.w / 2,
    touch.y + touch.h / 2,
    Math.max(touch.w, touch.h) / 2,
    0,
    2 * Math.PI
  );
  context.fill();
}

export function handleTouches(
  event: TouchEvent,
  context: CanvasRenderingContext2D
) {
  let simpleTouchEvent = simplifyTouchEvent(event);
  draw(simpleTouchEvent.touches[0]!, context);
}
