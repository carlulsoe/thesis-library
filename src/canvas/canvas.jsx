import React, { useEffect, useRef } from 'react';

export function Canvas({ setup: setup, width: width, height: height }) {
  const canvasRef = useRef(null);

  useEffect(() => {
    const canvas = canvasRef.current;
    const context = canvas.getContext('2d');
    setup(context);
  }, [setup]);

  return <canvas ref={canvasRef} width={width} height={height} />;
}
