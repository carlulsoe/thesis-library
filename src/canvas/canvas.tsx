import React, { useState } from 'react';
import { View, StyleSheet, type GestureResponderEvent } from 'react-native';
import Svg, { Path } from 'react-native-svg';
import {
  MenuProvider,
  Menu,
  MenuOptions,
  MenuTrigger,
  MenuOption,
} from 'react-native-popup-menu';

export const Canvas = () => {
  const [currentColor, setColor] = useState(Colors.Black);
  const [oldPaths, setOldPaths] = useState<IPath[]>([]);
  const [currentPath, setCurrentPath] = useState('');
  const addToPath = (s: string) => setCurrentPath(currentPath + s);
  const addCurrentToOldPaths = () =>
    setOldPaths([...oldPaths, { path: currentPath, color: currentColor }]);

  function handleStart(e: GestureResponderEvent) {
    let x = e.nativeEvent.touches[0]?.locationX;
    let y = e.nativeEvent.touches[0]?.locationY;
    addToPath(`M${x} ${y} `);
  }
  function handleMove(e: GestureResponderEvent) {
    let x = e.nativeEvent.touches[0]?.locationX;
    let y = e.nativeEvent.touches[0]?.locationY;
    addToPath(`L${x} ${y} `);
  }

  function handleEnd() {
    addCurrentToOldPaths();
    setCurrentPath('');
  }

  function clearCanvas() {
    setCurrentPath('');
    setOldPaths([]);
  }

  // @ts-ignore
  return (
    <MenuProvider>
      <View style={styles.outer}>
        <View
          style={styles.container}
          onTouchStart={handleStart}
          onTouchMove={handleMove}
          onTouchEnd={handleEnd}
        >
          <Svg width={1000} height={1000}>
            {oldPaths.map((path) => (
              <Path
                key={path.path}
                d={path.path}
                stroke={path.color}
                fillOpacity={0}
              />
            ))}
            <Path d={currentPath} stroke={currentColor} fillOpacity={0} />
          </Svg>
        </View>
        <View>
          <Menu>
            <MenuTrigger text={'Options'} style={styles.button} />
            <MenuOptions customStyles={optionsStyles}>
              <MenuOption
                onSelect={() => console.log(currentPath)}
                text={'Print path to console'}
              />
              <MenuOption
                onSelect={() => clearCanvas()}
                text={'Clear canvas'}
              />
              <MenuOption
                onSelect={() => setColor(Colors.Red)}
                text={'Red stroke'}
              />
              <MenuOption
                onSelect={() => setColor(Colors.Green)}
                text={'Green stroke'}
              />
              <MenuOption
                onSelect={() => setColor(Colors.Blue)}
                text={'Blue stroke'}
              />
              <MenuOption
                onSelect={() => setColor(Colors.Black)}
                text={'Black stroke'}
              />
            </MenuOptions>
          </Menu>
        </View>
      </View>
    </MenuProvider>
  );
};

const styles = StyleSheet.create({
  outer: {
    flex: 1,
  },
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    alignSelf: 'stretch',
  },
  button: {
    backgroundColor: '#2196F3',
    padding: 10,
    textAlign: 'center',
  },
});

const optionsStyles = {
  optionsContainer: {
    padding: 5,
  },
  optionWrapper: {
    margin: 5,
  },
  optionTouchable: {
    activeOpacity: 70,
  },
};

enum Colors {
  Red = '#FF0000',
  Green = '#00FF00',
  Blue = '#0000FF',
  Black = '#000000',
}

interface IPath {
  color: string;
  path: string;
}
