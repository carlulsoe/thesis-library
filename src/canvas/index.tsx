// @ts-ignore
export { Canvas } from './canvas';
export type { CanvasProps, CanvasSize } from './Canvas.types';
