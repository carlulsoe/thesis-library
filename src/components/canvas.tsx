import React, { useCallback, useContext, useState } from 'react';
import {
  type GestureResponderEvent,
  StyleSheet,
  useWindowDimensions,
  View,
} from 'react-native';
import Svg, { Path } from 'react-native-svg';
import {
  Menu,
  MenuOption,
  MenuOptions,
  MenuProvider,
  MenuTrigger,
} from 'react-native-popup-menu';
import { objEq, uniqueMerge } from '../extra/tools';
import { ConnectionContext } from '../connection/ConnectionContext';
import { useAutoUpdater } from '../connection/useAutoUpdater';

export const Canvas = () => {
  const { height, width } = useWindowDimensions();
  const [currentColor, setColor] = useState(Colors.Black);
  const [currentPath, setCurrentPath] = useState('');
  const [localPaths, setLocalPaths] = useState<IPath[]>([]);
  const [paths, setPaths] = useAutoUpdater('paths');
  const addToPath = (s: string) => setCurrentPath(currentPath + s);
  const addCurrentToOldPaths = () =>
    setLocalPaths([...localPaths, { path: currentPath, color: currentColor }]);
  const Context = useContext(ConnectionContext);

  function handleStart(e: GestureResponderEvent) {
    let x = e.nativeEvent.touches[0]?.pageX;
    let y = e.nativeEvent.touches[0]?.pageY;
    addToPath(`M${x} ${y} `);
  }

  function handleMove(e: GestureResponderEvent) {
    console.log('ah');
    let x = e.nativeEvent.touches[0]?.pageX;
    let y = e.nativeEvent.touches[0]?.pageY;
    addToPath(`L${x} ${y} `);
  }

  function handleEnd() {
    addCurrentToOldPaths();
    setCurrentPath('');
  }

  function clearCanvas() {
    setCurrentPath('');
    setLocalPaths([]);
  }

  const mergePaths = useCallback(() => {
    if (Context?.container == null) return;
    let remotePaths: IPath[] = paths ? JSON.parse(paths) : [];
    if (objEq(remotePaths, localPaths)) return;
    return uniqueMerge(remotePaths, localPaths);
  }, [Context?.container, localPaths, paths]);

  React.useEffect(() => {
    let allPaths = mergePaths();
    if (!allPaths) return;
    setLocalPaths(allPaths);
    setPaths(JSON.stringify(allPaths));
  }, [mergePaths, paths, setPaths]);

  // TODO: Error in height on canvas
  return (
    <MenuProvider>
      <View
        style={styles.container}
        onTouchStart={handleStart}
        onTouchMove={handleMove}
        onTouchEnd={handleEnd}
      >
        <Svg width={width} height={height}>
          {localPaths.map((path) => (
            <Path
              key={path.path}
              d={path.path}
              stroke={path.color}
              fillOpacity={0}
            />
          ))}
          <Path d={currentPath} stroke={currentColor} fillOpacity={0} />
        </Svg>
      </View>
      <View>
        <Menu>
          <MenuTrigger text={'Options'} style={styles.button} />
          <MenuOptions customStyles={optionsStyles}>
            <MenuOption
              onSelect={() => console.log(currentPath)}
              text={'Print path to console'}
            />
            <MenuOption
              onSelect={() => clearCanvas()}
              text={'Clear components'}
            />
            <MenuOption
              onSelect={() => setColor(Colors.Red)}
              text={'Red stroke'}
            />
            <MenuOption
              onSelect={() => setColor(Colors.Green)}
              text={'Green stroke'}
            />
            <MenuOption
              onSelect={() => setColor(Colors.Blue)}
              text={'Blue stroke'}
            />
            <MenuOption
              onSelect={() => setColor(Colors.Black)}
              text={'Black stroke'}
            />
            <MenuOption
              onSelect={() => setPaths(mergePaths())}
              text={'Update canvas'}
            />
          </MenuOptions>
        </Menu>
      </View>
    </MenuProvider>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    alignSelf: 'stretch',
  },
  button: {
    backgroundColor: '#2196F3',
    padding: 10,
    textAlign: 'center',
  },
});

const optionsStyles = {
  optionsContainer: {
    padding: 5,
  },
  optionWrapper: {
    margin: 5,
  },
  optionTouchable: {
    activeOpacity: 70,
  },
};

enum Colors {
  Red = '#FF0000',
  Green = '#00FF00',
  Blue = '#0000FF',
  Black = '#000000',
}

type IPath = {
  color: string;
  path: string;
};
