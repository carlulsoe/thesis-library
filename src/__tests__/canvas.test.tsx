import React from 'react';
import renderer from 'react-test-renderer';
import { Canvas } from '../canvas';

test('renders correctly', () => {
  const tree = renderer
    .create(
      <Canvas
        setup={(ctx) => {
          ctx.save();
        }}
        height={200}
        width={200}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
