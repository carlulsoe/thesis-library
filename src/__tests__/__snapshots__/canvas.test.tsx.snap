// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<canvas
  height={200}
  width={200}
/>
`;
